### PR TITLE
[pwm, dv] Add testplan for PWM

### PR DESCRIPTION
**Pull request by tunghoang290780**
Add pwm_testplan.md file for dv

Signed-off-by: Hoang Tung <Hoang.Tung@wdc.com>

### CODE REVIEW

**Comment by senelson7** [comment]
The descriptions for blink/heartbeat tests is a copy of the description for smoke test and should be modified.

File: hw/ip/pwm/data/pwm_testplan.hjson [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+      desc: '''
+            Checking ip operation at min/max bandwidth
+
+            Stimulus:
+              - Program timing registers (CLK_DIV, DC_RESN) to high/low values (slow/fast data rate),
+              - Program other required registers for pwm operation
+              - Start pwm channels
+
+            Checks:
+              -, Ensure the output pulses are correctly modulated for all channels
```
Comment (rasmus-madsen) [inline hw/ip/pwm/data/pwm_testplan.hjson:39]: remove comma

File: hw/ip/pwm/data/pwm_testplan.hjson [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+            '''
+      milestone: V2
+      tests: ["pwm_perf"]
+    }
+    {
+      name: blink
+      desc: '''
+            Smoke test accessing a major datapath within the pwm.
```
Comment (rasmus-madsen) [inline hw/ip/pwm/data/pwm_testplan.hjson:47]: remove punctuation

File: hw/ip/pwm/data/pwm_testplan.hjson [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+            Stimulus:
+              - Program pwm operation register (PWM_EN)
+              - Program configuration registers (CFG, INVERT)
+              - Enabke blink mode (BLINK_EN) and program DUTY_CYCLE register
+              - Program REGEN to lock (enable) register settings and start pwm function
+
+            Checks:
+              - Ensure the output pulses are correctly modulated for all channels,
```
Comment (rasmus-madsen) [inline hw/ip/pwm/data/pwm_testplan.hjson:56]: remove comma

File: hw/ip/pwm/data/pwm_testplan.hjson [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+              - Ensure blinking channel toggles by two duty cycle values programmed to DUTY_CYCLE register
+            '''
+      milestone: V1
+      tests: ["pwm_blink"]
+    }
+    {
+      name: hearbeat
+      desc: '''
+            Smoke test accessing a major datapath within the pwm.
```
Comment (rasmus-madsen) [inline hw/ip/pwm/data/pwm_testplan.hjson:65]: punctuation

File: hw/ip/pwm/data/pwm_testplan.hjson [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+              - Ensure output duty cycle linearly increments and decrements between two
+                blink-rate values programmed to BLINK_PARAM register
+            '''
+      milestone: V1
+      tests: ["pwm_hearbeat"]
+    }
+    {
+      name: stress_all
+      desc: '''
+            Combine above sequences in one test then randomly select for running.
```
Comment (rasmus-madsen) [inline hw/ip/pwm/data/pwm_testplan.hjson:84]: punctuation

**Review by rasmus-madsen — CHANGES_REQUESTED** [review]
@tunghoang290780 a few nits about punctuation
I think we generally don't use punctuation for one-liners, but at least you need to be consistent

**Comment by tunghoang290780** [comment]
@rasmus-madsen update PR to fix the issues

**Review by senelson7 — APPROVED** [review]
LGTM

File: hw/ip/pwm/data/pwm_testplan.hjson [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    }
+    {
+      name: blink
+      desc: '''
+            Test the blinking mode of pwm
+
+            Stimulus:
+              - Program pwm enable register (PWM_EN)
+              - Program configuration registers (CFG, INVERT)
+              - Enabke blink mode (BLINK_EN) and program DUTY_CYCLE register
```
Comment (weicaiyang) [inline hw/ip/pwm/data/pwm_testplan.hjson:52]: Enabke -> Enable

File: hw/ip/pwm/data/pwm_testplan.hjson [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+              - Program REGEN to lock (enable) register settings and start pwm function
+
+            Checks:
+              - Ensure the output pulses are correctly modulated for all channels
+              - Ensure blinking channel toggles by two duty cycle values programmed to DUTY_CYCLE register
+            '''
+      milestone: V1
```
Comment (weicaiyang) [inline hw/ip/pwm/data/pwm_testplan.hjson:59]: it this V2?

File: hw/ip/pwm/data/pwm_testplan.hjson [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+              - Program configuration registers (CFG, INVERT)
+              - Enable heartbeat mode (HTBT_EN) and programm BLINK_PARAM register
+              - Program REGEN to lock (enable) register settings and start pwm function
+
+            Checks:
+              - Ensure the output pulses are correctly modulated for all channels
+              - Ensure output duty cycle linearly increments and decrements between two
+                blink-rate values programmed to BLINK_PARAM register
+            '''
+      milestone: V1
```
Comment (weicaiyang) [inline hw/ip/pwm/data/pwm_testplan.hjson:78]: ditto, is this V2?

**Review by weicaiyang — APPROVED** [review]
LGTM with a few minor comments

File: hw/ip/pwm/data/pwm_testplan.hjson (hunk 7 above)
Comment (tunghoang290780) [reply hw/ip/pwm/data/pwm_testplan.hjson:59]: Ops, fixed it

File: hw/ip/pwm/data/pwm_testplan.hjson (hunk 6 above)
Comment (tunghoang290780) [reply hw/ip/pwm/data/pwm_testplan.hjson:52]: fixed

File: hw/ip/pwm/data/pwm_testplan.hjson (hunk 8 above)
Comment (tunghoang290780) [reply hw/ip/pwm/data/pwm_testplan.hjson:78]: should be V1

**Comment by tunghoang290780** [comment]
> The descriptions for blink/heartbeat tests is a copy of the description for smoke test and should be modified.

In the smoke test, the heartbeat and blink enable bits are disabled

**Comment by tunghoang290780** [comment]
@weicaiyang @rasmus-madsen update PR to address your comments

**Comment by weicaiyang** [comment]
> @weicaiyang @rasmus-madsen update PR to address your comments

LGTM, thanks

File: hw/ip/pwm/data/pwm_testplan.hjson [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+      milestone: V2
+      tests: ["pwm_heartbeat"]
+    }
+    {
+      name: clock_domain
+      desc: '''
+            TBD -Verify the function of pwm in both bus and core clock domains
+
+            Stimulus:
+              - Start pwm transaction in both bus and core clock domain
```
Comment (martin-lueker) [inline hw/ip/pwm/data/pwm_testplan.hjson:87]: I'm not sure what this means?  All transactions are started in the bus clock domain.   The register values are then fed through a CDC and executed in the core clock domain.

**Review by martin-lueker — CHANGES_REQUESTED** [review]
LGTM, though we should discuss the CDC tests a bit more.  I'm not understanding what you have in mind for this test.

File: hw/ip/pwm/data/pwm_testplan.hjson (hunk 9 above)
Comment (tunghoang290780) [reply hw/ip/pwm/data/pwm_testplan.hjson:87]: @martin-lueker I would change that to TBD and get back later. Could you put a waveform for CDC if possible?

**Review by rasmus-madsen — APPROVED** [review]
LGTM

**Comment by rasmus-madsen** [comment]
@martin-lueker I will let you merge this one - as you requested some changes

**Comment by tunghoang290780** [comment]
> @martin-lueker I will let you merge this one - as you requested some changes

@rasmus-madsen @martin-lueker I would like to merge this one (leave CDC test as TBD) since it allows me to move with DV which requires the testplan as a part of the build/sim flow

File: hw/ip/pwm/data/pwm_testplan.hjson (hunk 9 above)
Comment (martin-lueker) [reply hw/ip/pwm/data/pwm_testplan.hjson]: Okay.

**Review by martin-lueker — APPROVED** [review]

**Pull request merged**